lib/route/cls.c

```diff
@@ -364,6 +364,76 @@ void rtnl_cls_cache_set_tc_params(struct nl_cache *cache,
 	cache->c_iarg2 = parent;
 }
 
+/**
+ * Search classifier by interface index, parent and handle
+ * @arg cache           Classifier cache
+ * @arg ifindex         Interface index
+ * @arg parent          Parent
+ * @arg handle          Handle
+ *
+ * Searches a classifier cache previously allocated with rtnl_cls_alloc_cache()
+ * and searches for a classifier matching the interface index, parent
+ * and handle.
+ *
+ * The reference counter is incremented before returning the classifier,
+ * therefore the reference must be given back with rtnl_cls_put() after usage.
+ *
+ * @return Classifier or NULL if no match was found.
+ */
+struct rtnl_cls *rtnl_cls_get(struct nl_cache *cache, int ifindex, uint32_t parent,
+			      uint32_t handle)
+{
+	struct rtnl_cls *cls;
+
+	if (cache->c_ops != &rtnl_cls_ops)
+		return NULL;
+
+	nl_list_for_each_entry(cls, &cache->c_items, ce_list) {
+		if ((cls->c_parent == parent) && (cls->c_ifindex == ifindex)
+		    && (cls->c_handle == handle)) {
+			nl_object_get((struct nl_object *) cls);
+			return cls;
+		}
+	}
+
+	return NULL;
+}
+
+/**
+ * Search classifier by interface index, parent and priority
+ * @arg cache           Classifier cache
+ * @arg ifindex         Interface index
+ * @arg parent          Parent
+ * @arg prio            Priority
+ *
+ * Searches a classifier cache previously allocated with rtnl_cls_alloc_cache()
+ * and searches for a classifier matching the interface index, parent
+ * and prio.
+ *
+ * The reference counter is incremented before returning the classifier,
+ * therefore the reference must be given back with rtnl_cls_put() after usage.
+ *
+ * @return Classifier or NULL if no match was found.
+ */
+struct rtnl_cls *rtnl_cls_get_by_prio(struct nl_cache *cache, int ifindex,
+				      uint32_t parent, uint16_t prio)
```

Comment:
this patch adds two new API for things that are already possible (by the user implementing the functions themself).

Can you give some arguments why this is necessary and a good idea? Also, now we add two new functions, will this suffice? Is by-prio somehow specially interesting? Also, for a cache function called "get", I would expect that it can only find one possible value. Is the (ifindex,parent,prio) tuple in the cache unique? If not, shouldn't it be called "find_first_by_prio()"?

Also, the code comment should indicate that this that a search over all elements and is `O(n)`.

Reply:
It is indeed possible by the user to implement those functions, however, the beauty of any library is in the richness of its API functionality.

These two functions make things simpler for users, since:
- no need to implement them, they are there already
- use them in multiple applications, without having to duplicate code (this is primary reason for why I'm adding this patch)
- it seems to me that they are sort of _'natural extension'_ to what already is implemented

As for whether these two functions will suffice - I believe yes. Some classifiers allow to set unique handle when adding them (u32), others not (these end up having the same handle '0x1'). _Priority_ or _preference_ is possible to set for all classifiers: it must be unique, and is a good way to fetch wanted classifier from cache.

I agree with you that the naming of functions could be improved. How about:

- _rtnl_cls_find_by_handle()_: find cls in cache by unique handle (if this cls has a unique one)
- _rtnl_cls_find_by_prio()_: find cls in cache by priority (all classifiers have unique priority)

Reply:
makes sense to add the API.

I rename the function from "get" to "find".